src/wallet/rpcwallet.cpp

```diff
@@ -1155,19 +1157,36 @@ UniValue cleanwallettransactions(const UniValue& params, bool fHelp, const CPubK
         // get all locked utxos to relock them later.
         vector<COutPoint> vLockedUTXO;
         pwalletMain->ListLockedCoins(vLockedUTXO);
-        // unlock all coins so that the following call containes all utxos.
+        // unlock all coins so that the following call contains all utxos.
         pwalletMain->UnlockAllCoins();
-        // listunspent call... this gets us all the txids that are unspent, we search this list for the oldest tx,
-        vector<COutput> vecOutputs;
-        assert(pwalletMain != NULL);
-        pwalletMain->AvailableCoins(vecOutputs, false, NULL, true);
+
+        // this gets us all the txids that are unspent, we search for the oldest tx
         int32_t oldestTxDepth = 0;
-        BOOST_FOREACH(const COutput& out, vecOutputs)
+        for (map<uint256, CWalletTx>::const_iterator it = pwalletMain->mapWallet.begin(); it != pwalletMain->mapWallet.end(); ++it)
         {
-          if ( out.nDepth > oldestTxDepth )
-              oldestTxDepth = out.nDepth;
+            const uint256& wtxid = it->first;
+            const CWalletTx* pcoin = &(*it).second;
+
+            //Check if transaction will be final in the next block to be created.
+            if (!CheckFinalTx(*pcoin))
+                continue;
+
+            //orphaned coins get cleaned
+            int nDepth = pcoin->GetDepthInMainChain();
```

Comment:
This is a somewhat heavy call. A possible optimization is to temporarily store this data to make it unnecessary to call this again in the removal loop below. (scope creep, I know.)

Reply:
I see, I read `GetDepthInMainChainINTERNAL` and think I understand what you're saying. I would like to attempt to solve this issue but will need to work on it next week.

Reply:
If you do decide to tackle this, you may wish to do it in a separate PR. It is a performance fix, not a bug fix, and truly scope-creep. I have the bad habit of trying to fix everything when I should only fix 1 thing.

Reply:
It's all good! I believe I handled the other two things. I believe the original author had added in the unlocking to get a full list of transactions which as you said seems like it's not required. 

Builds on my system after these 2 changes, will try it out on one of my servers.